File: 2-hard/following_integer/following_integer.rb
def following_integer(number_string)
	number = number_string.chars.map {|char| char.to_i}.unshift 0
	last_digit_index = number.count - 1

	(last_digit_index - 1).downto(0) do |i|
		last_digit_index.downto(i) do |j|
			if number[j] > number[i]
				number[j], number[i] = number[i], number[j]
				higher_digits = number[0, i + 1]
				lower_digits = number[i + 1, number.count].sort
				result = higher_digits + lower_digits
				result.shift if result[0] == 0
				return result.join("")
			end
		end
	end

end

File.open(ARGV[0]).each_line do |line|
	number_string = line.strip
	puts following_integer(number_string)
end

File: 2-hard/following_integer/following_integer.php
<?php

function nextNumber($numberString)
{
    $number = str_split($numberString);
    for ($i = 0; $i < count($number); $i++) {
        $number[$i] = intval($number[$i]);
    }
    array_unshift($number, 0);
    $lastIndex = count($number) - 1;

    for ($i = $lastIndex - 1; $i >= 0; $i--) {
        for ($j = $lastIndex; $j > $i; $j--) {
            if ($number[$j] > $number[$i]) {
                //swap them
                $tmp = $number[$j];
                $number[$j] = $number[$i];
                $number[$i] = $tmp;

                //sort ascending all digits from $i + 1 to the end
                $lowerDigits = array_slice($number, $i + 1);
                sort($lowerDigits);
                $higherDigits = array_slice($number, 0, $i + 1);
                $number = array_merge($higherDigits, $lowerDigits);
                if ($number[0] == 0) {
                    array_shift($number);
                }

                return join("", $number);
            }
        }
    }

}

$fileName = $argv[1];
$lines = file($fileName);

foreach ($lines as $line) {
    echo nextNumber(trim($line))."\n";
}
